### PR TITLE
AP_Proximity: support multiple backends

### PR DESCRIPTION
**Pull request by rishabsingh3003**
This builds on top of @rmackay9 PR here: https://github.com/ArduPilot/ardupilot/pull/21222
Includes some last remaining fixes that needed to go in. 
Most details are given in the previous PR, but I have included the following changes:

1. Did the PRX_ to PRX1_ conversion 
2. Reduced the usage of "primary sensor" to report sensor data/status which was being used through the library. Now, in most places, we cycle through all available sensors. 
3. Created a new file called "AP_Proximity_Utils" to include utility functions and not clutter frontend too much. 
4. Pre-arm check now checks all sensors (not just the primary sensor)
5. I have increased the number of usable Proximity sensors to 3

### CODE REVIEW

**Comment by rishabsingh3003** [comment]
Screenshot from 2022-08-22 22-26-27

This is how pre arm checks are displayed now

**Comment by patrickpoirier51** [comment]
Tested good on Orange cube
2 cigbot looking at +45 and -45
1 lightware looking backward 

https://media.discordapp.net/attachments/836353372294086698/1011345169720807484/unknown.png

**Comment by rmackay9** [comment]
We could drop the Sub commit to do param conversion I suspect.  It's hard to imagine there are any sub users of the proximity sensor.

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 1]
```diff
@@ -195,90 +200,46 @@ float AP_Proximity::distance_min() const
...
-    if (!valid_instance(primary_instance)) {
-        return false;
-    }
-    // get closest object from backend
-    return drivers[primary_instance]->get_closest_object(angle_deg, distance);
+    return boundary.get_closest_object(angle_deg, distance);
 }
 
-// get number of objects, used for non-GPS avoidance
+// get number of objects, angle and distance - used for non-GPS avoidance
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:344]: this looks like a copy-paste error in the comment.  the get_object_count() method doesn't return angle and distance.

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 2]
```diff
@@ -392,7 +392,14 @@ bool AP_Proximity::get_upward_distance(uint8_t instance, float &distance) const
 
 bool AP_Proximity::get_upward_distance(float &distance) const
 {
-    return get_upward_distance(primary_instance, distance);
+    // get upward distance from backend
+    for (uint8_t i=0; i<num_instances; i++) {
+        // return first good upward distance
+        if (get_upward_distance(primary_instance, distance)) {
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:398]: "primary_instance" should probably be "i"?

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 3]
```diff
@@ -243,10 +243,36 @@ AP_Proximity::Status AP_Proximity::get_status(uint8_t instance) const
...
+    return Status::Good;
+}
+
+// prearm checks
+bool AP_Proximity::prearm_healthy(char *failure_msg, const uint8_t failure_msg_len) const
+{
+    for (uint8_t i=0; i<num_instances; i++) {
+        switch (get_instance_status(i)) {
+        case Status::NoData:
+            hal.util->snprintf(failure_msg, failure_msg_len, "PRX %X: No Data", i + 1);
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:263]: nitpick: %X is for uppercase hexadecimal numbers it seems.  Perhaps change this to "PRX%d:".  Save some flash by removing the space too.

**Comment by rmackay9** [comment]
I'm not a huge fan of moving some methods into a Utility file.  I just think it'll make them harder to find.  I'll leave this to @rishabsingh3003 to decide though.

**Comment by rmackay9** [comment]
We could merge these commits together:

- AP_Proximity: extend how long shorter distance may remain
- AP_Proximity: boundary uses shortest dist from multiple backends

-- and these two:

- AP_Proximity: add potential support for up to 4 proxmity sensors
- AP_Proximity: move params to separate class

**Comment by rmackay9** [comment]
I've looked through this now and besides the comments I've made  i think it looks pretty good.

**Review by rmackay9 — APPROVED** [review]
LGTM but let's get input from others on the weekly dev call.

**Pull request merged**